Implementation Netsleuth algorithm

Retzoh opened the pull request:
The Netsleuth implementation as presented in the article "Spotting Culprits in Epidemics, How many and Which ones ?".
 There are 2 troubling things :

The first concerns the method to select the number of newly infected nodes among the frontier in the ripple-simulation (line 150) : with the proposed method (n = floor( probability_getting_infected \* ( 1 + | frontier | / transmission_probability ) ) ) based on a supposedly scale-invariance through 1 / transmission_probability, the total frontier is systematically selected, so the actual calculation is pointless.

The second concerns the heart of the algorithm : the method for the selection of the optimal sources (line 83) seems more than doubtful to me : you should take the node with the highest coordinate in the eigenvector associated to the smallest eigenvalue of the laplacian matrix of the graph. But the eigenvalues are all positive, and 0 is an obvious eigenvalue with [1, ..., 1] for eigenvector so the highest coordinate makes no sense. On the other hand, each eigenvector for the 0 value corresponds to a connected component of the graph. In the article, the method is resumed in a sentence 3 times (IV C), each time differently. I took the second one, the one that seemed most clearly expressed to me. Perhaps there is something I did not understand in the 2 others.

A technical problem I had is by computing the eigenvectors : the numpy and scipy.linalg methods to compute them make it through numerical approximation and give completely incoherent values for the vector considering the 0-eigenvalue. I found a method from Matrix, with analytical calculation and exact values but the complexity is to check.

Commit:
Implementation Netsleuth algorithm

The Netsleuth implementation as presented in the article "Spotting Culprits in Epidemics, How many and Which ones ?".
 There are 2 troubling things :

The first concerns the method to select the number of newly infected nodes among the frontier in the ripple-simulation (line 150) : with the proposed method (n = floor( probability_getting_infected * ( 1 + | frontier | / transmission_probability ) ) ) based on a supposedly scale-invariance through 1 / transmission_probability, the total frontier is systematically selected, so the actual calculation is pointless.

The second concerns the heart of the algorithm : the method for the selection of the optimal sources (line 83) seems more than doubtful to me : you should take the node with the highest coordinate in the eigenvector associated to the smallest eigenvalue of the laplacian matrix of the graph. But the eigenvalues are all positive, and 0 is an obvious eigenvalue with [1, ..., 1] for eigenvector so the highest coordinate makes no sense. On the other hand, each eigenvector for the 0 value corresponds to a connected component of the graph. In the article, the method is resumed in a sentence 3 times (IV C), each time differently. I took the second one, the one that seemed most clearly expressed to me. Perhaps there is something I did not understand in the 2 others.

A technical problem I had is by computing the eigenvectors : the numpy and scipy.linalg methods to compute them make it through numerical approximation and give completely incoherent values for the vector considering the 0-eigenvalue. I found a method from Matrix, with analytical calculation and exact values but the complexity is to check.

Temigo commented:
Concerning the second problem : it is the smallest eigenvalue of the **submatrix** of the laplacian matrix, corresponding to the infected graph. So in my opinion 0 won't be an obvious eigenvalue as you said ?

Temigo commented:
Concerning the first problem : I suspect there is a mistake in the paper. We should read `n = floor( probability_getting_infected * ( 1 + frontier  ) )`. Otherwise ( 1 + | frontier | / transmission_probability ) will always be greater than |frontier| since we divide it by a probability (0<=p<=1). It is also more coherent with the statement regarding the Binomial (parameters n = f_d, k = m_d).

Note : there is also a mistake in the first mention of this Binomial law. Since the parameters are as I said, we should read (m_d in f_d) and not (k in f_d) (definition of probability of infection, III.B).

Retzoh commented:
For the second one, I agree but I was working with the sub-matrix in the first one.
I am making a second pull-request : I forgot the case of no-connected graphs.

Commit:
Correction working for non-connected graphs



The Netsleuth implementation as presented in the article "Spotting Culprits in Epidemics, How many and Which ones ?".
There are 2 troubling things :

The first concerns the method to select the number of newly infected nodes among the frontier in the ripple-simulation (line 150) : with the proposed method (n = floor( probability_getting_infected * ( 1 + | frontier | / transmission_probability ) ) ) based on a supposedly scale-invariance through 1 / transmission_probability, the total frontier is systematically selected, so the actual calculation is pointless.

The second concerns the heart of the algorithm : the method for the selection of the optimal sources (line 83) seems more than doubtful to me : you should take the node with the highest coordinate in the eigenvector associated to the smallest eigenvalue of the laplacian matrix of the graph. But the eigenvalues are all positive, and 0 is an obvious eigenvalue with [1, ..., 1] for eigenvector so the highest coordinate makes no sense. On the other hand, each eigenvector for the 0 value corresponds to a connected component of the graph. In the article, the method is resumed in a sentence 3 times (IV C), each time differently. I took the second one, the one that seemed most clearly expressed to me. Perhaps there is something I did not understand in the 2 others.

A technical problem I had is by computing the eigenvectors : the numpy and scipy.linalg methods to compute them make it through numerical approximation and give completely incoherent values for the vector considering the 0-eigenvalue. I found a method from Matrix, with analytical calculation and exact values but the complexity is to check.

Retzoh commented:
And for the second one : it means that their assumption of scale-invariance is wrong...

Temigo commented:
Right, I forgot their assumption of scale-invariance.

For the second issue, do you mean that you get zero as an eigenvalue each time, even if using the submatrix of the laplacian ?

Temigo commented:
Actually I agree with their scale invariance in time in 1/beta and accordingly we might understand their sentence _a set of size f_d is equivalent to a set of size f_d/beta_ as _the next frontier will be a set of size f_d/beta_. But it doesn't explain the formula, which could be understandable only for the next frontier set.

I guess they might have mixed up a bit, what does it give if we forget their 1/beta invariance ?

Retzoh commented:
0 is an eigenvalue for each Laplacian matrix and the submatrix considered here is the laplacian matrix of the infected graph.

Retzoh commented:
And I don't agree with you for the size of the next frontier : if each node has a single uninfected neighbour, not in common with any of the others, not in the frontier then the frontier size is invariant whatever the value of beta.

Temigo commented:
Nope, I don't think it is the laplacian matrix of the infected graph : The diagonal values are the degrees of the nodes in the graph and it won't be equal to their degree in the infected subgraph (at least for the nodes on the frontier of the infection), if I am not mistaken...

Retzoh commented:
In table II at the beginning : La is the laplacian matrix of the graph Gi. But you might have a clue

Temigo commented:
They don't say that : they write _submatrix (size NI x NI) of L(G) corresponding to the infected graph GI_ which is slightly different.

Temigo commented:
But you're right for the size of the next frontier. It's really odd. I propose we notify this strange formula to the MAP group... and go along with the formula without beta.

Retzoh commented:
I see where I was wrong. I will take it in account and see if it gets better.

Retzoh commented:
I have taken care of the matrix problem with better results : I was indeed taking a wrong matrix. the new code will be coming when I have finished testing it but until now I have encountered a new problem : for computing the eigenvectors, I can either use the linalg.eigh() method from scipy which computes numerically and gives really  wrong eigenvectors half of the time, or the eigenvects() method from sympy.Matrix, which computes analytically but takes far too much time, already with small graphs in my tests. As I have a big step in execution time when increasing the graph size, I fear there is something I forgot when I wrote it, causing a loop, so I will be looking for it but if you know a goof method for eigenvectors calculations, I am interested.

Temigo commented:
For SciPy : I think you should use the `linalg.eig` function instead of the `linalg.eigh` (which is solely for hermitian or symmetric matrices, but L_A isn't symmetric as we said).

Temigo commented on algorithm_netsleuth.py:
```diff
@@ -7,4 +7,247 @@
+
+        i_laplacian_matrix = nx.laplacian_matrix(i_graph)
+        # Diagonalizing the laplacian matrix
+        val, vect = linalg.eigh(i_laplacian_matrix.todense()) # The eigh function gives wrong eigenvects for eigenval 0
```
Why do you use `todense()` method ?

Temigo commented:
I haven't any clue for SymPy. It shouldn't be that slow ! Can you commit and pull your changes so that I can have a look ?

Commit:
Update algorithm_netsleuth.py

Commit:
Merge pull request #1 from Retzoh/Retzoh-patch-1

Implementation Netsleuth algorithm

Retzoh commented:
L_A is actually symmetric : only the diagonal terms differ from the laplacian matrix.
The default structure for a sparse matrix given by the nx method is some compressed structure that is not directly usable so I had to use the todense() method to get the matrix under a conventional form, understandable by scipy.linalg and sympy.Matrix.

For the speed : the Matrix.eigenvects() method really gets slow from 25 infected nodes, at which it is almost instantaneous with the scipy method so it is quite problematic. I may be that I don't have enough memory on my computer...

I left the two methods in the code, one being commented.

Retzoh commented:
You also have to consider the fact that my plotting function is taking quite some time. When I am saying that the Matrix method takes too long time, it means that I abort the calculations after some 20 minutes to get back to work.

Retzoh commented:
I found another library working fast and with good results : scipy.sparse.linalg.eigs() I am loading it up.

Commit:
Final Netsleuth algorithm

Found a good method for the eigenvector problem. The scale-invariance has been taken back.
Now open for comparing to the others !

Commit:
Better plotting

An example with two sources

Commit:
Merge pull request #2 from Retzoh/Retzoh-patch-1

Better plotting
